### ISSUE TITLE
Generalize Publish API

### ISSUE THREAD

**Issue by esoterra** [bytecodealliance/registry#84]
Expand the publish api to use the `content_needed` mechanism so clients can automatically discover where/how to upload content.

**Comment by peterhuene** [comment]
This can possibly be closed with the recent updates to the REST API implementation in the client/server.

However, what the API currently does not support is a way for the server to tell the client that it should upload the content to a location that isn't the server itself; clients can still inform the server of known locations for the content already, if supported by the server (`warg-server` currently only supports direct content uploads; I personally don't think it's a good idea for a registry to connect to user-supplied hosts to be able to download and check content against policy).

I think we need to really consider why informing the client as to where it should upload its content is needed in the API.

At a surface level, it feels like it isn't buying us much when you take into consideration that the registry will want to stream new content _anyway_ to validate it against content policy.

So whether or not the registry configures some S3 bucket with a special access token only for a particular upload, informs the client of the location (plus access token), client uploads the content and then notifies the server the upload is complete, the server must then turn around and download the content from S3 to check it against policy, then revoke the access token so the bucket is never writable again (_ick_ to even temporarily writable buckets from public IPs, access token or not).

If a general purpose registry can't scale to being the go-between for the client and a content storage service like S3 (i.e. client uploads the content to the registry, the registry incrementally scans it, and writes the stream out to a never-writable-by-anything-but-the-registry S3 bucket as it goes), then it probably shouldn't be in the "general purpose registry" business.

**Comment by lann** [comment]
I agree that any particular registry would likely only have one supported upload method, and that for a BA-operated registry that may be "upload to the registry directly". IIRC we were advised that "large cloud providers" wouldn't like that approach and would want to push directly to object stores, which matches what I have seen in some large cloud provder service designs.

Separately, several BA members (Fermyon included) would like to back packages with OCI instead of HTTP; there are certainly other ways to implement that feature but that was a design consideration.

**Comment by Kylebrown9** [comment]
As per what Lann said, there are reasons people may want to direct content uploads to different places. So, we need to specify it (including in the OpenAPI) and should exercise them in our reference implementation regardless of whether certain registry instances won't need it.

**Comment by calvinrp** [comment]
@macovedj  and I are looking into a PR for the upload flow API endpoints and CLI support. Thinking about blob store signed URL policies, multipart uploads and hash integrity checks. As well as, OCI Registry push.
